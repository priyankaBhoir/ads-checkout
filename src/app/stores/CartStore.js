'use strict'

import dispatcher from '../dispatcher/Dispatcher';
import {EventEmitter} from 'events';
import { createStore } from './Store';
import Constants from '../constants/Constants';
import calculatePrice from '../utils/RuleEngine';

const {
	ACTION_CONSTANT,
	EVENT_CONSTANT
} = Constants;


//======================= functions to update store ================
let _cart = {
	products: {},
	customerName: "default"
};

function updateAds (data) {
	_cart = data;
}

function add(ad, quantity) {
	_cart.products[ad.id] = _cart.products[ad.id] ? 
		Object.assign({}, _cart.products[ad.id], {quantity: _cart.products[ad.id].quantity + quantity}) :
		Object.assign({}, ad, {quantity: quantity});

}
function updateCustomer (name) {
	_cart.customerName = name;
}

function resetCart() {
	_cart = {
		products: {},
		customerName: "default"
	};
}

//======================= Store ===============================
const CartStore = createStore ({
	currentState () {
		return _cart;
	},

	getTotalNumberOfItems() {
		return Object.keys(_cart.products).reduce((total, id) => total + _cart.products[id].quantity, 0);
	},

	getTotalPrice() {
		return {
			actualPrice: calculatePrice(Object.assign({}, _cart, {customerName: "default"})),	//calculate witthout discount
			discountPrice: calculatePrice(_cart) //calculate with discount
		}
	}
});

//======================= Store's Listerner ===========================
CartStore.dispatchToken = dispatcher.register (action => {

	switch (action.type) {
		case ACTION_CONSTANT.ADD_TO_CART:
			add(action.ad, action.quantity);
			CartStore.emitChange(EVENT_CONSTANT.ITEM_ADDED_TO_CART);
			break;
		case ACTION_CONSTANT.UPDATE_CUSTOMER_NAME:
			updateCustomer(action.name);
			CartStore.emitChange(EVENT_CONSTANT.RECACULATE_CART);
			break;
		case ACTION_CONSTANT.INIT_CART: 
			resetCart();
			break;

	}
});

export default CartStore;