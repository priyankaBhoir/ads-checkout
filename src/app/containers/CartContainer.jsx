import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CartStore from '../stores/CartStore';
import {addToCart, setCustomerName} from '../actions/CartActions';
import Constant from '../constants/Constants'
import {
  Link
} from 'react-router-dom';

export default class CartContainer extends Component {
	constructor(props) {
		super(props);
		this.state = {
			items: CartStore.currentState().products,
			totalPrices: CartStore.getTotalPrice()
		}
	}

	componentDidMount() {
		CartStore.addChangeListner(Constant.EVENT_CONSTANT.RECACULATE_CART, this.recalculate)
	}

	componentWillUnmount() {
		CartStore.removeChangeListner(Constant.EVENT_CONSTANT.RECACULATE_CART, this.recalculate)
	}

	recalculate = () => {
		this.setState({
			totalPrices: CartStore.getTotalPrice()
		})
	}

	setCustomerName = () => {		
		setCustomerName(this.refs.customerName.value)
	}

	render() {
		return (
			<div>
				<div className="section"> 
					{Object.keys(this.state.items).map(id => {
						return (<div className="row"> 
								<span className="col m4"> {this.state.items[id].title }</span> 
								<span className="col m4"> $ {this.state.items[id].price }</span> 
								<span className="col m4"> {this.state.items[id].quantity }</span> 
							</div>)
					})} 
				</div>
				<div class="divider"/> 
				<div className="section">
					<div>Your total: <b> $ {this.state.totalPrices.actualPrice} </b></div>
					{this.state.totalPrices.discountPrice !== this.state.totalPrices.actualPrice && 
						<div> Your total After Discount: <b> $ {this.state.totalPrices.discountPrice} </b></div>
					}
				</div>
				<div class="divider"/> 
				<div className="section">
					<span> Enter your name to check special discounts on your purchase </span>
					<div className="row">
						<input className="col m7" type="text" ref="customerName" />
						<button className="col m4" onClick={this.setCustomerName} > Save </button>
					</div>
				</div>
				<div className="center link"> <span class="btn grey lighten-2"> <Link to="/">continue shopping...</Link></span></div>
			</div>
		)
	}
}
