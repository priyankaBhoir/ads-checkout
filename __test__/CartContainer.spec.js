import React from 'react'
import { shallow, mount, unmount } from 'enzyme';
import renderer from 'react-test-renderer';
import {MemoryRouter } from 'react-router-dom'
import CartContainer from '../src/app/containers/CartContainer';

import Dispatcher from '../src/app/dispatcher/Dispatcher';
import AppConstants from '../src/app/constants/Constants';
import CartStore from '../src/app/stores/CartStore';
import CartAction from '../src/app/actions/CartActions';


jest.mock('react-router-dom');


const mountWithRouter = node => mount(<MemoryRouter>{node}</MemoryRouter>);

// Snapshot for Home React Component
describe('CartContainer --- Snapshot',()=>{
    it('+++capturing Snapshot of Container', () => {
        const renderedValue =  renderer.create(<CartContainer />).toJSON()
        expect(renderedValue).toMatchSnapshot();

    });
});

///*******************************************************************************************************
describe('Shallow Render REACT [SMART] Container :: CartContainer',()=>{
    let wrapper
    beforeEach(()=>{
      if(wrapper) wrapper.unmount();
       wrapper = mountWithRouter(<CartContainer/>);     
    })
    

    it('+++ Smart component render even if nothinng is in cart', () => {
      wrapper = mountWithRouter(<CartContainer />)
      expect(wrapper.length).toEqual(1)
    });

    //Does not pass, because it is enclosed in router.
    //@todo: https://github.com/ReactTraining/react-router/issues/5059
    /*it('+++ Smart component render, link to shopping should be present', () => {
      wrapper = mountWithRouter(<CartContainer />)
      console.log(wrapper.props().children);
      expect(wrapper.props().children.find('.link').length).toEqual(1)
    });*/

   
    
});
