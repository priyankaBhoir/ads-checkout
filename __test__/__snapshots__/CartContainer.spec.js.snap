// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartContainer --- Snapshot +++capturing Snapshot of Container 1`] = `
<div>
  <div
    className="section"
  />
  <div
    className="divider"
  />
  <div
    className="section"
  >
    <div>
      Your total: 
      <b>
         $ 
        0
         
      </b>
    </div>
  </div>
  <div
    className="divider"
  />
  <div
    className="section"
  >
    <span>
       Enter your name to check special discounts on your purchase 
    </span>
    <div
      className="row"
    >
      <input
        className="col m7"
        type="text"
      />
      <button
        className="col m4"
        onClick={[Function]}
      >
         Save 
      </button>
    </div>
  </div>
  <div
    className="center link"
  >
     
    <span
      className="btn grey lighten-2"
    >
       
    </span>
  </div>
</div>
`;
